source/agora/test/Ledger.d

```diff
@@ -71,3 +83,79 @@ unittest
     // higher index than available => return nothing
     assert(node_1.getBlocksFrom(1000, 10).length == 0);
 }
+
+/// test catch-up phase during booting
+unittest
+{
+    import core.thread;
+    import std.algorithm;
+    import std.conv;
+    import std.format;
+
+    const NodeCount = 4;
+    auto network = makeTestNetwork!TestNetworkManager(NetworkTopology.Simple, NodeCount);
+
+    auto nodes = network.apis.values;
+    auto node_1 = nodes[0];
+    auto gen_key_pair = getGenesisKeyPair();
+    auto gen_block = getGenesisBlock();
+
+    auto txes = getChainedTransactions(gen_block.tx, 100, gen_key_pair);
+    txes.each!(tx => node_1.putTransaction(tx));
+
+    assert(node_1.getBlockHeight() == 100);
+
+    foreach (empty_node; nodes[1 .. $])
+    {
+        assert(empty_node.getBlockHeight() == 0);
+    }
+
+    // now start the network, let other nodes catch-up the latest blocks
+    network.start();
+    assert(network.getDiscoveredNodes().length == NodeCount);
+
+    auto attempts = 80;  // wait up to 80*100 msecs (8 seconds)
+    while (attempts--)
+    {
+        if (allHeightsEqual(nodes, 100))
+            return;
+
+        // let them do catch-up after boot
+        Thread.sleep(100.msecs);
+    }
+}
+
+/// test catch-up phase after initial booting (periodic catch-up)
+unittest
+{
+    import core.thread;
+    import std.algorithm;
+    import std.format;
```

Comment:
Those 2 imports are not necessary anymore

Reply:
I need `each` still.